src/coordtest/src/lib.rs

```diff
@@ -0,0 +1,335 @@
+                                for r in result {
+                                    match r {
+                                        ExecuteResponse::SendingRows(rows) => match rows.await {
+                                            PeekResponse::Rows(rows) => {
+                                                for row in rows {
+                                                    for col in row.iter() {
+                                                        if col != Datum::True {
+                                                            failed = Err(anyhow!("datum != true"));
+                                                        }
+                                                    }
+                                                }
+                                            }
+                                            r => {
+                                                failed = Err(anyhow!("{:?}", r));
+                                            }
+                                        },
+                                        _ => panic!("expected SendingRows"),
+                                    };
+                                }
+                            }
+                            Err(err) => {
+                                failed = Err(err);
+                            }
+                        };
+                        match failed {
+                            Ok(_) => {
+                                break;
+                            }
+                            Err(err) => {
+                                if start.elapsed() > Duration::from_secs(5) {
+                                    panic!("{}", err);
+                                }
+                                // Bump the timestamp. This is necessary because sources ingest at varying
+                                // rates and we need to allow sinces to move forward so we can see new data.
+                                let mut ts = ct.timestamp.lock().unwrap();
+                                *ts += 1;
+                            }
+                        }
+                    }
+                    "".into()
+                }
+                "update-upper" => {
+                    let catalog = ct.make_catalog().await;
+                    let mut updates = vec![];
+                    for line in tc.input.lines() {
+                        let mut line = line.split_whitespace();
+                        let id = catalog.get(line.next().unwrap());
+                        let ts: Timestamp = line.next().unwrap().parse().unwrap();
+                        assert!(line.next().is_none());
+                        // A ts <= 1 won't do anything.
+                        assert!(ts > 1);
+                        let upper = ct.uppers.entry(id).or_insert(0);
+                        let mut batch: ChangeBatch<Timestamp> = ChangeBatch::new_from(*upper, -1);
+                        *upper = ts;
```

Comment:
it feels like we maybe want to check that ts >= upper here?

Reply:
done